### PR TITLE
Unit tests for deployment event handler

### PR DESCRIPTION
**Pull request by hks93**
Work in  Progress PR : 
Code not sanitized -dev in progress

### CODE REVIEW

File: jazz_deployments-event-handler/package.json [hunk 1]
```diff
@@ -9,13 +9,18 @@
     "test": "mocha"
   },
   "dependencies": {
+    "lodash": "*",
```
Comment (svidhani) [inline jazz_deployments-event-handler/package.json]: is this required?

File: jazz_deployments-event-handler/index.js [hunk 2]
```diff
@@ -114,7 +114,8 @@ var processEventRecord = (record, configData, authToken) => {
 	});
 };
 
-var checkForInterestedEvents = (encodedPayload, sequenceNumber, configData) => {
+function checkForInterestedEvents(encodedPayload, sequenceNumber, configData)  {
+	console.log("in the actual function 2");
```
Comment (svidhani) [inline jazz_deployments-event-handler/index.js]: remove this

File: jazz_deployments-event-handler/index.js [hunk 3]
```diff
@@ -366,7 +367,7 @@ module.exports = {
 	getTokenRequest: getTokenRequest,
 	getAuthResponse: getAuthResponse,
 	handleError: handleError,
-	checkForInterestedEvents: checkForInterestedEvents,
+	checkForInterestedEvents,
```
Comment (svidhani) [inline jazz_deployments-event-handler/index.js]: i guess all the functions will have to be updated not just checkForInterestedEvents

File: jazz_deployments-event-handler/test/test.js [hunk 4]
```diff
@@ -1,9 +1,327 @@
...
+    let responseObject = {
+      statusCode: 200,
+      body: {
+        data: {
+          message: message
+        }
+      }
+    };
+    
+    var checkForInterestedEvents = sinon.stub(index,"checkForInterestedEvents").withArgs("a","b","c").resolves("somthing");
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: please be consistent (use let or var as the case maybe)

File: jazz_deployments-event-handler/test/test.js [hunk 5]
```diff
@@ -1,9 +1,327 @@
...
+    }
+    sandbox = sinon.createSandbox();
+  })
+  afterEach(()=>{
+    sandbox.restore();
+    if(reqStub){
+      reqStub.restore();
+    }
+  })
+  it.only("should call processEvent for intrested events", () => {
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: don't forget to remove .only

File: jazz_deployments-event-handler/test/test.js [hunk 6]
```diff
@@ -1,9 +1,327 @@
...
+    let responseObject = {
+      statusCode: 200,
+      body: {
+        data: {
+          message: message
+        }
+      }
+    };
+    reqStub = sinon.stub(request, "Request", (obj) => {
+      console.log("stub is called ")
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: replace this with sinon.assert.calledOnce.
add this to all tests wherever you are calling the stub.

File: jazz_deployments-event-handler/test/test.js [hunk 7]
```diff
@@ -1,9 +1,327 @@
...
+})
+describe("getDeploymentPayload", () => {
+  var svcContext 
+  beforeEach(() => {
+     svcContext = {
+      "service_type": "api",
+      "branch": "",
+      "runtime": "nodejs",
+      "domain": "jazztest",
+      "iam_role": "arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1",
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 8]
```diff
@@ -1,9 +1,327 @@
...
+      "environment": "",
+      "region": "us-east-1",
+      "message": "input validation starts",
+      "created_by": "serverless@t-mobile.com"
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 9]
```diff
@@ -1,9 +1,327 @@
...
+          S: 'NOT_SERVICE_DEPLOYMENT'
+        },
+        USERNAME: {
+          S: 'serverless@t-mobile.com'
+        },
+        EVENT_TIMESTAMP: {
+          S: '2018-05-16T12:12:41:083'
+        },
+        SERVICE_CONTEXT: {
+          S: '{"service_type":"api","branch":"","runtime":"nodejs","domain":"jazztest","iam_role":"arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1","environment":"","region":"us-east-1","message":"input validation starts","created_by":"serverless@t-mobile.com"}'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 10]
```diff
@@ -1,9 +1,327 @@
...
+          S: '09ed3279-c8b9-e360-2a78-4e1ed093e6a7'
+        },
+        EVENT_STATUS: {
+          S: 'STARTED'
+        },
+        EVENT_TYPE: {
+          S: 'NOT_SERVICE_DEPLOYMENT'
+        },
+        USERNAME: {
+          S: 'serverless@t-mobile.com'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 11]
```diff
@@ -1,9 +1,327 @@
...
+        },
+        EVENT_TIMESTAMP: {
+          S: '2018-05-16T12:12:41:083'
+        },
+        SERVICE_CONTEXT: {
+          S: '{"service_type":"api","branch":"","runtime":"nodejs","domain":"jazztest","iam_role":"arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1","environment":"","region":"us-east-1","message":"input validation starts","created_by":"serverless@t-mobile.com"}'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 12]
```diff
@@ -1,9 +1,327 @@
...
+          S: '09ed3279-c8b9-e360-2a78-4e1ed093e6a7'
+        },
+        EVENT_STATUS: {
+          S: 'STARTED'
+        },
+        EVENT_TYPE: {
+          S: 'NOT_SERVICE_DEPLOYMENT'
+        },
+        USERNAME: {
+          S: 'serverless@t-mobile.com'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 13]
```diff
@@ -1,9 +1,327 @@
...
+        },
+        EVENT_TIMESTAMP: {
+          S: '2018-05-16T12:12:41:083'
+        },
+        SERVICE_CONTEXT: {
+          S: '{"service_type":"api","branch":"","runtime":"nodejs","domain":"jazztest","iam_role":"arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1","environment":"","region":"us-east-1","message":"input validation starts","created_by":"serverless@t-mobile.com"}'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 14]
```diff
@@ -1,9 +1,327 @@
...
+          S: '09ed3279-c8b9-e360-2a78-4e1ed093e6a7'
+        },
+        EVENT_STATUS: {
+          S: 'STARTED'
+        },
+        EVENT_TYPE: {
+          S: 'NOT_SERVICE_DEPLOYMENT'
+        },
+        USERNAME: {
+          S: 'serverless@t-mobile.com'
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 15]
```diff
@@ -1,9 +1,327 @@
...
+      statusCode: 401,
+      body: {}
+    }
+    index.getAuthResponse(result).catch((err) => {
+      expect(err.message).to.eq('Invalid token response from API');
+    })
+  })
+
+})
+describe("checkforIntrestedEvents", () => {
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: nitpick: fix typo `Interested`

File: jazz_deployments-event-handler/test/test.js [hunk 16]
```diff
@@ -1,9 +1,327 @@
...
+        "username": configData.SERVICE_USER,
+        "password": configData.TOKEN_CREDS
+      },
+      rejectUnauthorized: false,
+      transform: (body, response, resolveWithFullResponse) => {
+        return response;
+      }
+    };
+  })
+  it("should give return auth token when called with valid paramentes", () => {
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: nitpick: `parameters`

File: jazz_deployments-event-handler/test/test.js [hunk 17]
```diff
@@ -1,9 +1,327 @@
...
+var context = {
+  "callbackWaitsForEmptyEventLoop": true,
+  "logGroupName": "/aws/lambda/gitlab180515-jazz-services-handler-prod",
+  "logStreamName": "2018/05/16/[$LATEST]2beb5567c06b42dabe390268cfca8427",
+  "functionName": "gitlab180515-jazz-services-handler-prod",
+  "memoryLimitInMB": "256",
+  "functionVersion": "$LATEST",
+  "invokeid": "bb61d7b3-d439-4efe-9c6d-4ea78d7349d9",
+  "awsRequestId": "bb61d7b3-d439-4efe-9c6d-4ea78d7349d9",
+  "invokedFunctionArn": "arn:aws:lambda:us-east-1:192006145812:function:gitlab180515-jazz-services-handler-prod"
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 18]
```diff
@@ -1,9 +1,327 @@
...
+    "eventID": "shardId-000000000000:49584481860528260622422554690105735408360629197594427394",
+    "eventName": "aws:kinesis:record",
+    "invokeIdentityArn": "arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1",
+    "awsRegion": "us-east-1",
+    "eventSourceARN": "arn:aws:kinesis:us-east-1:192006145812:stream/gitlab180515-events-hub-prod"
+  }]
+}
+var context = {
+  "callbackWaitsForEmptyEventLoop": true,
+  "logGroupName": "/aws/lambda/gitlab180515-jazz-services-handler-prod",
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 19]
```diff
@@ -1,9 +1,327 @@
...
+  "logStreamName": "2018/05/16/[$LATEST]2beb5567c06b42dabe390268cfca8427",
+  "functionName": "gitlab180515-jazz-services-handler-prod",
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 20]
```diff
@@ -1,9 +1,327 @@
...
+      "approximateArrivalTimestamp": 1526472764.125
+    },
+    "eventSource": "aws:kinesis",
+    "eventVersion": "1.0",
+    "eventID": "shardId-000000000000:49584481860528260622422554690105735408360629197594427394",
+    "eventName": "aws:kinesis:record",
+    "invokeIdentityArn": "arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1",
+    "awsRegion": "us-east-1",
+    "eventSourceARN": "arn:aws:kinesis:us-east-1:192006145812:stream/gitlab180515-events-hub-prod"
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 21]
```diff
@@ -1,9 +1,327 @@
...
+      "approximateArrivalTimestamp": 1526472764.125
+    },
+    "eventSource": "aws:kinesis",
+    "eventVersion": "1.0",
+    "eventID": "shardId-000000000000:49584481860528260622422554690105735408360629197594427394",
+    "eventName": "aws:kinesis:record",
+    "invokeIdentityArn": "arn:aws:iam::192006145812:role/gitlab180515_lambda2_basic_execution_1",
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: unit tests don't require actual data, can you please scrub it asap?

File: jazz_deployments-event-handler/test/test.js [hunk 22]
```diff
@@ -1,9 +1,327 @@
...
+  })
+
+})
+describe("checkforIntrestedEvents", () => {
+  it("should return object with paramenter interested_event set to true", () => {
+    var record = event.Records[0];
+    var sequenceNumber = record.kinesis.sequenceNumber;
+    var encodedPayload = record.kinesis.data;
+    index.checkForInterestedEvents(encodedPayload, sequenceNumber, configData).then((res) => {
+      // console.log(res);
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: remove this

File: jazz_deployments-event-handler/test/test.js [hunk 23]
```diff
@@ -1,9 +1,736 @@
...
+    if (procesEverntRecordStub) {
+      procesEverntRecordStub.restore();
+    }
+  })
+  it("should resolve all for success scenario from processEventRecord",()=>{
+    procesEverntRecordStub = sinon.stub(index, "processEventRecord").resolves({
+      "status":"succesfully processed Event Rcord"
+    })
+    index.processEventRecords(event,configData,"temp_auth").then((obj)=>{
+      for(let i=0;i<obj.length;i++){expect(obj[i].status).to.eq("succesfully processed Event Rcord")}
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: put this in separate lines and also do sinon.assert.calledOnce on stubs everywhere

File: jazz_deployments-event-handler/test/test.js [hunk 24]
```diff
@@ -1,9 +1,736 @@
...
+      }
+    }
+  })
+  afterEach(() => {
+    if (reqStub) {
+      reqStub.restore();
+    }
+  })
+  it("should call getDeployments for true case scenarios", () => {
+    var temp = {
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: i guess this can go in beforeEach

Comment (hks93) [reply jazz_deployments-event-handler/test/test.js]: done

File: jazz_deployments-event-handler/index.js [hunk 25]
```diff
@@ -86,35 +83,35 @@ var processEventRecords = (event, configData, authToken) => {
 	});
 };
 
-var processEventRecord = (record, configData, authToken) => {
+function processEventRecord (record, configData, authToken){
```
Comment (svidhani) [inline jazz_deployments-event-handler/index.js]: probably for another PR(refactor) but these event related functions can be grouped in a separate module

File: jazz_deployments-event-handler/test/test.js [hunk 26]
```diff
@@ -1,9 +1,731 @@
...
+  afterEach(() => {
+    if (reqStub) {
+      reqStub.restore();
+    }
+  })
+  it("should call processEvents with SvcPayload", () => {
+    var procesRequestStub = sinon.stub(index, "procesRequest").resolves({
+      x: 1
+    })
+    index.processCreateEvent(payload.Item, configData, "tempAuth").then(() => {
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: what about a test for catch block

File: jazz_deployments-event-handler/test/test.js [hunk 27]
```diff
@@ -1,9 +1,731 @@
...
+      reqStub.restore();
+    }
+  })
+  it("should call getDeployments for true case scenarios", () => {
+    
+    var getDeploymentsStub = sinon.stub(index, "getDeployments").resolves(temp)
+    var updateDeploymentsStub = sinon.stub(index, "updateDeployments").resolves(temp)
+    index.processUpdateEvent(payload.Item, configData, "tempAuth").then(() => {
+      sinon.assert.calledOnce(getDeploymentsStub)
+      getDeploymentsStub.restore()
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: missing semi-colon here and other places in the code

File: jazz_deployments-event-handler/index.js [hunk 28]
```diff
@@ -362,23 +359,24 @@ var handleFailedEvents = function (id, failure_message, payload, failure_code) {
...
-	processEventRecord: processEventRecord,
-	processEvent: processEvent,
-	getDeploymentPayload: getDeploymentPayload,
-	getSvcPayload: getSvcPayload,
-	procesRequest: procesRequest,
-	processCreateEvent: processCreateEvent,
-	processUpdateEvent: processUpdateEvent,
-	getDeployments: getDeployments,
-	updateDeployments: updateDeployments
+ const factory= {
```
Comment (suryajak) [inline jazz_deployments-event-handler/index.js]: rename this to exportable. Also reduce the number of export functions here.

File: jazz_deployments-event-handler/index.js (hunk 25 above)
Comment (hks93) [reply jazz_deployments-event-handler/index.js]: WILL TAKE IT UP AS PART OF THE NEW STORY INTENDED TO REDUCE THE EXPORTED MODULES

File: jazz_deployments-event-handler/test/test.js [hunk 29]
```diff
@@ -1,9 +1,740 @@
...
+  it("should call processEvents with SvcPayload", () => {
+    var procesRequestStub = sinon.stub(index, "procesRequest").resolves({
+      x: 1
+    })
+    index.processCreateEvent(payload.Item, configData, "tempAuth").then(() => {
+      sinon.assert.calledOnce(procesRequestStub)
+      procesRequestStub.restore()
+    })
+  });
+  it("should call processEvents with SvcPayload", () => {
```
Comment (svidhani) [inline jazz_deployments-event-handler/test/test.js]: different description that the previous one

File: jazz_deployments-event-handler/test/test.js [hunk 30]
```diff
@@ -1,9 +1,740 @@
...
+    "eventSource": "aws:kinesis",
+    "eventVersion": "1.0",
+    "eventID": "shardId-000000000000:49584481860528260622422554690105735408360629197594427394",
+    "eventName": "aws:kinesis:record",
+    "invokeIdentityArn": "arn:aws:iam::12345678:role/gitlabtest10001_test01",
+    "awsRegion": "us-east-1",
+    "eventSourceARN": "arn:aws:kinesis:us-east-1:100000002:teststream/gitlab10001-events-hub-prod"
+  }]
+};
+var context = {
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: Unit tests missing for the main handler function

File: jazz_deployments-event-handler/test/test.js [hunk 31]
```diff
@@ -1,9 +1,740 @@
...
+  "logGroupName": "/aws/lambda/gitlab1234567-test-name",
+  "logStreamName": "2018/05/16/temp_test0000012910",
+  "functionName": "gitlab001-test-services-handler-prod",
+  "memoryLimitInMB": "256",
+  "functionVersion": "$LATEST",
+  "invokeid": "00001-test-000001",
+  "awsRequestId": "00001-test-000001",
+  "invokedFunctionArn": "arn:aws:lambda:us-east-1:100000001:function:gitlab001-test-services-handler-prod"
+};
+var configData = configObj(context);
```
Comment (suryajak) [inline jazz_deployments-event-handler/test/test.js]: Unit tests missing for failure scenarios. For example, I don't see any of the unit tests exercise the code path where `handleFailedEvents` will be called.

File: jazz_deployments-event-handler/index.js [hunk 32]
```diff
@@ -8,42 +8,39 @@ const _ = require("lodash");
 const request = require("request");
 const rp = require('request-promise-native');
 const Uuid = require("uuid/v4");
-
```
[the remark this answers is not in the document]
Comment (hks93) [reply jazz_deployments-event-handler/index.js]: done

File: jazz_deployments-event-handler/test/test.js (hunk 30 above)
Comment (hks93) [reply jazz_deployments-event-handler/test/test.js]: Done

File: jazz_deployments-event-handler/test/test.js (hunk 31 above)
Comment (hks93) [reply jazz_deployments-event-handler/test/test.js]: Need clarification - handleFailedEvents has only 5 lines of code and is intended to push the failed event to an array 
didn't find a condition to be tested

Comment (suryajak) [reply jazz_deployments-event-handler/test/test.js]: Setup a scenario where you multiple events to process. Stub `processRequest` where you make this `reject` once and `resolve` all the other times. have multiple eve and make one of them.

File: jazz_deployments-event-handler/index.js [hunk 33]
```diff
@@ -223,32 +237,32 @@ var getSvcPayload = (method, payload, apiEndpoint, authToken) => {
 	return svcPayload;
 };
 
-var procesRequest = (svcPayload) => {
+function procesRequest(svcPayload){
```
Comment (suryajak) [inline jazz_deployments-event-handler/index.js]: typo in function name: `proces**s**Request `

File: jazz_deployments-event-handler/test/test.js (hunk 31 above)
Comment (hks93) [reply jazz_deployments-event-handler/test/test.js]: Added - line 184

**Pull request merged**